### PR TITLE
[RLlib] Migrating DDPG to PolicyV2

### PR DESCRIPTION
**Pull request by charlesjsun**
## Why are these changes needed?

Migrating DDPG to PolicyV2

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/ddpg/ddpg_tf_policy.py [hunk 1]
```diff
@@ -454,61 +88,362 @@ def update_target_fn(tau):
...
-    mixins=[
-        TargetNetworkMixin,
-        ActorCriticOptimizerMixin,
-        ComputeTDErrorMixin,
-    ],
-)
+# We need this builder function because we want to share the same
+# custom logics between TF1 dynamic and TF2 eager policies.
+def get_ddpg_tf_policy(base: Type[Union[DynamicTFPolicyV2, EagerTFPolicyV2]]) -> Type:
+    """Construct a SimpleQTFPolicy inheriting either dynamic or eager base policies.
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/ddpg/ddpg_tf_policy.py:104]: Although linter does not complain: We commonly place a black line between sections of docstrings.

Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/ddpg/ddpg_tf_policy.py:104]: SimpleQTFPolicy

File: rllib/algorithms/ddpg/ddpg_tf_policy.py [hunk 2]
```diff
@@ -454,61 +88,362 @@ def update_target_fn(tau):
...
+
+        @override(base)
+        def make_model(self) -> ModelV2:
+            return make_ddpg_models(self)
+
+        @override(base)
+        def optimizer(
+            self,
+        ) -> Union[
+            "tf.keras.optimizers.Optimizer", List["tf.keras.optimizers.Optimizer"]
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/ddpg/ddpg_tf_policy.py:154]: Only `List["tf.keras.optimizers.Optimizer"]` should suffice here!

File: rllib/algorithms/sac/sac_tf_policy.py [hunk 3]
```diff
@@ -670,6 +669,65 @@ def setup_early_mixins(
     ActorCriticOptimizerMixin.__init__(policy, config)
 
 
+class ComputeTDErrorMixin:
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/sac/sac_tf_policy.py:672]: Let's put a TODO comment here that this TDErrorMixIn should be unified with the DDPG one, once we sub-class the SAC policy to avoid redundancy here!

File: rllib/algorithms/sac/sac_tf_policy.py [hunk 4]
```diff
@@ -670,6 +669,65 @@ def setup_early_mixins(
...
+                    PRIO_WEIGHTS: tf.convert_to_tensor(importance_weights),
+                },
+            )
+            # `self.td_error` is set in loss_fn.
+            return self.td_error
+
+        self.compute_td_error = compute_td_error
+
+
+class TargetNetworkMixin:
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/sac/sac_tf_policy.py:699]: Same here - let's mark this as a TODO!

File: rllib/utils/spaces/simplex.py [hunk 5]
```diff
@@ -23,7 +23,6 @@ class Simplex(gym.Space):
 
     def __init__(self, shape, concentration=None, dtype=np.float32):
         assert type(shape) in [tuple, list]
-        assert len(shape) == 2, ""
```
Comment (ArturNiederfahrenhorst) [inline rllib/utils/spaces/simplex.py:26]: Thanks for this fix!

File: rllib/policy/dynamic_tf_policy_v2.py [hunk 6]
```diff
@@ -251,15 +251,12 @@ def compute_gradients_fn(
     @OverrideToImplementCustomLogic
     def apply_gradients_fn(
         self,
-        policy: Policy,
```
Comment (ArturNiederfahrenhorst) [inline rllib/policy/dynamic_tf_policy_v2.py:254]: Why was this here before?

File: rllib/algorithms/td3/tests/test_td3.py [hunk 7]
```diff
@@ -54,6 +54,7 @@ def test_td3_exploration_and_with_random_prerun(self):
 
         # Test against all frameworks.
         for _ in framework_iterator(config, with_eager_tracing=True):
+            config.seed = 42
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/td3/tests/test_td3.py:57]: I think we should avoid seeding wherever possible. Maybe others differ!

File: rllib/algorithms/ddpg/ddpg_torch_policy.py [hunk 8]
```diff
@@ -28,287 +28,360 @@
...
-    # update critic net `policy_delay` time(s).
-    if policy.global_step % policy.config["policy_delay"] == 0:
-        policy._actor_optimizer.step()
+    @override(TorchPolicyV2)
+    def loss(
+        self,
+        model: ModelV2,
+        dist_class: Type[TorchDistributionWrapper],
+        train_batch: SampleBatch,
+    ) -> Union[TensorType, List[TensorType]]:
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/ddpg/ddpg_torch_policy.py:215]: Here again, I think this should only return a List of Tensors.

**Review by ArturNiederfahrenhorst — COMMENTED** [review]
Only some nits, thanks for your work @charlesjsun !

File: rllib/policy/dynamic_tf_policy_v2.py (hunk 6 above)
Comment (charlesjsun) [reply rllib/policy/dynamic_tf_policy_v2.py:254]: Because in the policy builder api `policy` need to be passed in as the first argument, but `self` takes care of it in PolicyV2, and this is probably leftover from copy-paste and was never found since apply_gradients_fn was never used in a PolicyV2 algorithm until now

Comment (ArturNiederfahrenhorst) [reply rllib/policy/dynamic_tf_policy_v2.py:254]: Thanks!

**Review by ArturNiederfahrenhorst — APPROVED** [review]
Awesome, thanks!

File: rllib/algorithms/td3/tests/test_td3.py (hunk 7 above)
Comment (sven1977) [reply rllib/algorithms/td3/tests/test_td3.py]: Yeah, I agree. This seed here has been around for quite some time, though (I think this line was just move here).
So, ok to leave for now.

**Comment by sven1977** [comment]
Thanks for this PR @charlesjsun ! and for your review @ArturNiederfahrenhorst !
We are making great progress toward a cleaner Policy API.
